### PR TITLE
Make Skopeo tests work with either Docker or Podman runtimes

### PR DESCRIPTION
**Pull request by linguini-dev**
PR #20009 broke some tests which did not have podman pre-installed but only docker. This PR will make Skopeo tests run with either Podman or Docker depending on the use-case.

- Related test failure: https://openqa.suse.de/tests/15398447#step/skopeo/56
- Needles: N/A
- Verification run: http://dirtman.qe.prg2.suse.org/tests/41

### CODE REVIEW

**Review by grisu48 — APPROVED** [review]
LGTM

**Review by pdostal — APPROVED** [review]

**Review by ricardobranco777 — CHANGES_REQUESTED** [review]
I don't like disabling this test for docker.  There's a reason we created this test:
- https://bugzilla.suse.com/show_bug.cgi?id=1219563
- https://progress.opensuse.org/issues/154933

I believe the best compromise is to avoid the podman commands `if check_var("CONTAINER_RUNTIMES", "docker")`

**Comment by ggardet** [comment]
Fails on opensuse-Tumbleweed-DVD-aarch64-Build20240910-containers_host_docker@aarch64 https://openqa.opensuse.org/tests/4470706#step/skopeo/56

**Comment by linguini-dev** [comment]
> I don't like disabling this test for docker. There's a reason we created this test:
> 
>     * https://bugzilla.suse.com/show_bug.cgi?id=1219563
> 
>     * https://progress.opensuse.org/issues/154933
> 
> 
> I believe the best compromise is to avoid the podman commands `if check_var("CONTAINER_RUNTIMES", "docker")`

Although I agree, Skopeo does not need a container runtime to run. I.e. it doesn't matter if podman or docker is installed.So if we want to be 100% technically correct, the test should run regardless of either. It just needs to be in the container context.

TL;DR; the issue is misleading and not entirely relevant to the Skopeo test.

**Comment by ricardobranco777** [comment]
> Although I agree, Skopeo does not need a container runtime to run. I.e. it doesn't matter if podman or docker is installed.So if we want to be 100% technically correct, the test should run regardless of either. 

That's why I think it should be run on _both_. 

> TL;DR; the issue is misleading and not entirely relevant to the Skopeo test.

The issue that motivated the creation of this test - apart from not having skopeo tests at that time - was integration of Docker & Skopeo.

**Comment by ricardobranco777** [comment]
> > Although I agree, Skopeo does not need a container runtime to run. I.e. it doesn't matter if podman or docker is installed.So if we want to be 100% technically correct, the test should run regardless of either.
> 
> That's why I think it should be run on _both_.
> 
> > TL;DR; the issue is misleading and not entirely relevant to the Skopeo test.
> 
> The issue that motivated the creation of this test - apart from not having skopeo tests at that time - was integration of Docker & Skopeo.

With this PR we're losing needed test coverage instead of increasing it.

**Comment by linguini-dev** [comment]
> > > Although I agree, Skopeo does not need a container runtime to run. I.e. it doesn't matter if podman or docker is installed.So if we want to be 100% technically correct, the test should run regardless of either.
> > 
> > 
> > That's why I think it should be run on _both_.
> > > TL;DR; the issue is misleading and not entirely relevant to the Skopeo test.
> > 
> > 
> > The issue that motivated the creation of this test - apart from not having skopeo tests at that time - was integration of Docker & Skopeo.
> 
> With this PR we're losing needed test coverage instead of increasing it.

Skopeo does not need a container runtime, which means that even if we run the Skopeo test with neither docker nor podman, the test will work as intended. Skopeo however does connect to an image registry (either the SUSE one or the Docker one), which is still part of the test.

What I mean is that we should run the Skopeo test as a separate test, and avoid relying on the container runtime variables. Isn't it better to have tests that verify what we need instead of having more tests which do not give us actual information?

**Comment by grisu48** [comment]
@ricardobranco777 Please see https://progress.opensuse.org/issues/163553#note-10, it was decided that there API compatibility between docker and skopeo in this case is a rather rare corner case that we cannot support and/or test.

This was the result of a lengthy discussion with the container engine maintainers.

**Review by ricardobranco777 — APPROVED** [review]
LGTM! Thanks!

**Comment by ggardet** [comment]
Squash commits maybe?

File: tests/containers/skopeo.pm [hunk 1]
```diff
@@ -94,7 +94,7 @@ sub run {
     assert_script_run("diff -urN $dir1 $dir2", fail_message => 'Copied images are not identical.');
 
     ######### Spin-up an instance of the latest Registry
-    assert_script_run("podman run --rm -d -p 5050:5000 --name skopeo-registry registry.suse.com/suse/registry:latest",
+    assert_script_run("$self->{runtime} run --rm -d -p 5050:5000 --name skopeo-registry registry.suse.com/suse/registry:latest",
```
Comment (ricardobranco777) [inline tests/containers/skopeo.pm:97]:
```suggestion
    assert_script_run("$runtime run --rm -d -p 5050:5000 --name skopeo-registry registry.suse.com/suse/registry:latest",
```

File: tests/containers/skopeo.pm [hunk 2]
```diff
@@ -126,12 +126,13 @@ sub run {
...
 sub cleanup {
+    my ($self) = @_;
     record_info('Cleanup', 'Delete copied image directories');
     script_run "rm -rf $workdir";
 
     record_info('Cleanup Registry', 'Remove local image Registry');
-    script_run "podman stop skopeo-registry";
-    script_run "podman rm -vf skopeo-registry";
+    script_run "$self->{runtime} stop skopeo-registry";
+    script_run "$self->{runtime} rm -vf skopeo-registry";
```
Comment (ricardobranco777) [inline tests/containers/skopeo.pm:135]:
```suggestion
    script_run "$runtime stop skopeo-registry";
    script_run "$runtime rm -vf skopeo-registry";
```

File: tests/containers/skopeo.pm [hunk 3]
```diff
@@ -19,10 +19,10 @@ use containers::common qw(install_packages);
 my $workdir = '/tmp/test';
 
 sub run {
-    my ($self, $args) = @_;
-
     # Required packages
     my @packages = qw(skopeo jq);
+    my ($self, $args) = @_;
+    my $runtime = $args->{runtime};
```
Comment (ricardobranco777) [inline tests/containers/skopeo.pm:25]: I would make this variable global so it can be reused by the `cleanup()` function.

Comment (linguini-dev) [reply tests/containers/skopeo.pm:25]: tbh this is the first thing that I tried and it did not work - the cleanup function failed because `$runtime` was replaced with an empty string. I am not sure why :\

**Comment by pdostal** [comment]
Related ticket: [poo#166781](https://progress.opensuse.org/issues/166781)
